exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/scan/project/ReaderSchemaOrchestrator.java

```diff
@@ -68,10 +71,11 @@ public void setBatchSize(int size) {
 
   @VisibleForTesting
   public ResultSetLoader makeTableLoader(TupleMetadata readerSchema) {
-    return makeTableLoader(scanOrchestrator.scanProj.context(), readerSchema);
+    return makeTableLoader(scanOrchestrator.scanProj.context(), readerSchema, -1);
   }
 
-  public ResultSetLoader makeTableLoader(CustomErrorContext errorContext, TupleMetadata readerSchema) {
+  public ResultSetLoader makeTableLoader(CustomErrorContext errorContext,
+      TupleMetadata readerSchema, long localLimit) {
```

Comment:
Sorry for the question, how to understand the variable name : localLimit?
If need 50 rows (for two bacth), and A batch has only 40 rows, so detect the B batch only need to fetch (50 - 40 =) 10 rows, then 10 rows is the localLimit?

Reply:
Good question. Let's start at the top. The user specifies a query limit, say, `LIMIT 50000`. Drill is distributed, so the limit must be pushed to each of the scan operators. Let's say there are 2, each running in a different thread. I hope the Calcite planner pushes the full limit to each, so each scan gets a limit of 50,000. (Reason: we can't know ahead of time if either scan can produce that many rows, so we have to pass the full amount in case the other scan produces no rows.)

I just realized one other issue: the limit can be pushed into the scan only if there is no `WHERE` clause in the query. It is actually rather unusual to *not* have a `WHERE` clause, so this whole limit-pushdown feature targets a rather simple, and unusual use case.

Any, assume we bet past the above issues, and we pass that 50K limit down into the scan. That is the "global limit" for the scan. (Or, maybe a better term term is "scan limit" since the limit is not global to all scans, just this one.)

Now, every scan can process any number of readers. Let's say this scan has 3 readers. Each reader has its own "local limit". Let's see how that works. The first reader gets the full limit of 50K rows. Let's say the reader actually returns 10K rows. In that case, the second reader gets a local limit of 40K rows. And so on.

Once the local limit goes to zero, we skip any remaining readers: they can't give us any new rows, so no need to even start those readers.

So, this function accepts the per-reader ("local") scan limit

Reply:
Got it.